File: core/src/test/java/com/zeyad/rxredux/core/redux/UserListVM.java
package com.zeyad.rxredux.core.redux;

import com.zeyad.gadapter.ItemInfo;
import com.zeyad.usecases.api.IDataService;
import com.zeyad.usecases.requests.GetRequest;
import com.zeyad.usecases.requests.PostRequest;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.functions.BiFunction;
import io.reactivex.functions.Function;

/**
 * @author zeyad on 11/1/16.
 */
public class UserListVM extends BaseViewModel<UserListState> {

    private final IDataService dataUseCase;

    public UserListVM(IDataService dataUseCase) {
        this.dataUseCase = dataUseCase;
    }

    @Override
    public StateReducer<UserListState> stateReducer() {
        return (newResult, event, currentStateBundle) -> {
            List<User> users;
            if (currentStateBundle == null || currentStateBundle.getUsers() == null)
                users = new ArrayList<>();
            else users = Observable.fromIterable(currentStateBundle.getUsers())
                    .map(ItemInfo::<User>getData).toList().blockingGet();
            List<User> searchList = new ArrayList<>();
            switch (event) {
                case "GetPaginatedUsersEvent":
                    users.addAll((List) newResult);
                    break;
                case "SearchUsersEvent":
                    searchList.addAll((List) newResult);
                    break;
                case "DeleteUsersEvent":
                    users = Observable.fromIterable(users)
                            .filter(user -> !((List) newResult).contains(user.getLogin()))
                            .distinct().toList().blockingGet();
                    break;
                default:
                    break;
            }
            return UserListState.builder()
                    .users(users)
                    .searchList(searchList)
                    .lastId(users.get(users.size() - 1).getId())
                    .build();
        };
    }

    @Override
    public Function<BaseEvent, Flowable<?>> mapEventsToActions() {
        return event -> {
            Flowable action = Flowable.empty();
            if (event instanceof GetPaginatedUsersEvent) {
                action = getUsers(((GetPaginatedUsersEvent) event).getPayLoad());
            } else if (event instanceof DeleteUsersEvent) {
                action = deleteCollection(((DeleteUsersEvent) event).getPayLoad());
            } else if (event instanceof SearchUsersEvent) {
                action = search(((SearchUsersEvent) event).getPayLoad());
            }
            return action;
        };
    }

    private Flowable<List<User>> getUsers(long lastId) {
        return lastId == 0 ?
                dataUseCase.getListOffLineFirst(new GetRequest.Builder(User.class, true)
                        .url(String.format("", lastId))
                        .cache(User.LOGIN)
                        .build())
                : dataUseCase.getList(new GetRequest.Builder(User.class, true)
                .url(String.format("", lastId)).build());
    }

    private Flowable<List<User>> search(String query) {
        return dataUseCase.<User>queryDisk(realm -> realm.where(User.class).beginsWith(User.LOGIN, query))
                .zipWith(dataUseCase.<User>getObject(new GetRequest.Builder(User.class, false)
                                .url(String.format("", query)).build())
                                .onErrorReturnItem(new User()).filter(user -> user.getId() != 0)
                                .map(user -> user != null ? Collections.singletonList(user)
                                        : Collections.emptyList()),
                        (BiFunction<List<User>, List<User>, List<User>>) (users, singleton) -> {
                            users.addAll(singleton);
                            return new ArrayList<>(new HashSet<>(users));
                        });
    }

    private Flowable<List<String>> deleteCollection(List<String> selectedItemsIds) {
        return dataUseCase.deleteCollectionByIds(new PostRequest.Builder(User.class, true).payLoad(selectedItemsIds)
                .idColumnName(User.LOGIN, String.class).cache().build()).map(o -> selectedItemsIds);
    }
}

File: core/src/test/java/com/zeyad/rxredux/core/redux/UserListVM.kt
package com.zeyad.rxredux.core.redux

import com.zeyad.gadapter.ItemInfo
import com.zeyad.rxredux.core.BaseEvent
import com.zeyad.rxredux.core.viewmodel.BaseViewModel
import com.zeyad.usecases.api.IDataService
import com.zeyad.usecases.requests.GetRequest
import com.zeyad.usecases.requests.PostRequest
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.functions.Function

class UserListVM(private val dataUseCase: IDataService) : BaseViewModel<UserListState>() {

    override fun stateReducer(): (newResult: Any, event: BaseEvent<*>, currentStateBundle: UserListState) -> UserListState {
        return { newResult, event, currentStateBundle ->
            var users: MutableList<User> = if (currentStateBundle.getUsers() == null)
                mutableListOf()
            else
                Observable.fromIterable<ItemInfo>(currentStateBundle.getUsers())
                        .map { it.getData<User>() }
                        .toList().blockingGet()
            val searchList = mutableListOf<User>()
            when (event) {
                is GetPaginatedUsersEvent -> users.addAll(newResult as List<User>)
                is SearchUsersEvent -> searchList.addAll(newResult as List<User>)
                is DeleteUsersEvent -> users = Observable.fromIterable(users)
                        .filter { user -> !(newResult as List<*>).contains(user.getLogin()) }
                        .distinct().toList().blockingGet()
                else -> {
                }
            }
            UserListState.builder()
                    .users(users)
                    .searchList(searchList)
                    .lastId(users[users.size - 1].getId().toLong())
                    .build()
        }
    }

    override fun mapEventsToActions(): Function<BaseEvent<*>, Flowable<*>> {
        return Function { event ->
            when (event) {
                is GetPaginatedUsersEvent -> getUsers(event.getPayLoad())
                is DeleteUsersEvent -> deleteCollection(event.getPayLoad())
//                is SearchUsersEvent -> search(stateEvent.payLoad)
                else -> {
                    throw IllegalArgumentException()
                }
            }
        }
    }

    private fun getUsers(lastId: Long): Flowable<List<User>> {
        return if (lastId == 0L)
            dataUseCase.getListOffLineFirst(GetRequest.Builder(User::class.java, true)
                    .url(String.format("user/%s", lastId))
                    .cache(User.LOGIN)
                    .build())
        else
            dataUseCase.getList(GetRequest.Builder(User::class.java, true)
                    .url(String.format("", lastId)).build())
    }

//    private fun search(query: String): Flowable<List<User>> {
//        return dataUseCase.queryDisk<User> { realm -> realm.where(User::class.java).beginsWith(User.LOGIN, query) }
//                .zipWith(dataUseCase.getObject<User>(GetRequest.Builder(User::class.java, false)
//                        .url(String.format("user/%s", query)).build())
//                        .onErrorReturnItem(User()).filter { user -> user.getId() != 0 }
//                        .map<Any> { user -> listOf(user) },
//                        { users, singleton ->
//                            users.addAll(singleton)
//                            ArrayList<User>(HashSet<User>(users))
//                        } as BiFunction<List<User>, List<User>, List<User>>)
//    }

    private fun deleteCollection(selectedItemsIds: List<String>): Flowable<List<String>> {
        return dataUseCase.deleteCollectionByIds<Any>(PostRequest.Builder(User::class.java, true).payLoad(selectedItemsIds)
                .idColumnName(User.LOGIN, String::class.java).cache().build()).map { o -> selectedItemsIds }
    }
}
